### DISCUSSION TITLE
Parameter file XML include not looking in the distribution

### DISCUSSION THREAD

**Discussion by jones-michael-s** [automaticanalysis/automaticanalysis#248] [General]
Finally getting around to organizing our parameter files by using the include feature (rather than customizing a full file). For example:

```
<?xml version="1.0" encoding="utf-8"?>
<aap xmlns:xi="http://www.w3.org/2001/XInclude">
    <xi:include href="aap_parameters_defaults.xml" parse="xml"/>
    <local>
	<directory_conventions>
        <spmdir desc="Path(s) to SPM software (>SPM12 r6470 required)" ui="dir">/Applications/spm12</spmdir>		
	</directory_conventions>
    </local>
</aap>
```
However, the xi:include fails unless I copy `aap_default_parameters.xml` to my working directory.  This is odd, because Matlab knows `aap_default_parameters.xml` lives in the repo (i.e., it's returned by `which('aap_default_parameters.xml') `).

Is this the intended behavior?

**Comment by tiborauer** [comment]
Do you store your parameterset file(s) within aa (i.e. next to `aap_parameters_defaults.xml`?
If not, then you need to provide a full path to `aap_parameters_defaults.xml`. That is also how the user-specific `aap_parameters_user.xml` includes the site-specific parameterset.

**Comment by tiborauer** [comment]
So, yes it is somewhat expected behaviour. `XML include` does not rely on MATLAB because it is not a MATLAB tool but an XML namespace (https://www.w3.org/TR/xinclude).

**Comment by jones-michael-s** [comment]
Okay, just verified that either storing my parameter file in /aa_parametersets fixes the problem and so does specifying a fullpath in the xml include. Neither are optimal solutions, but the latter is the lesser evil as we shouldn't require a user to modify the repo just to use AA.

**Comment by tiborauer** [comment]
You do not have to modify the repo when storing the parameterset (under a distinct name) in aa_parametersets. You do not have to commit them or worry about losing them during push/pull.
I think the whole idea of having aa_parametersets is to store parametersets there. Alternatively, we could just add (a few of) them to the examples; which also sounds plausible to me.

**Comment by jones-michael-s** [comment]
I don't mean modifying in the sense of push/pull.

Simply adding a file to the repo is "modifying the repo". 

User customization files don't belong in the code distribution. It's poor design.

**Comment by tiborauer** [comment]
Do you suggest, then, adding these parametersets to the examples?

**Comment by jones-michael-s** [comment]
I think there's already an adequate set of example parameter files in the distribution. IMO, what would really help is expanding the documentation. You've made a good start on the AA wiki parameterset section, but there's some critical details missing (like specifying a full path to the parameter file to be included, unless you put your parameter file in the distribution, which we now all agree is a bad thing :-)

**Comment by tiborauer** [comment]
Sorry. I meant MOVING (the subset of) the existing parametersets to the examples and omit the aa_parametersets folder.

I agree on expanding the documentation. @hhentschke is going to go through the ["Getting Started" section of the website](https://automaticanalysis.github.io/gettingstarted/), but feel free to contribute to it via PRs.

**Comment by jones-michael-s** [comment]
> Sorry. I meant MOVING (the subset of) the existing parametersets to the examples and omit the aa_parametersets folder.

Hmm. When I hear "examples" I think example _scripts_. But I don't feel strongly either way. 

> I agree on expanding the documentation.

I'm currently finishing up an AA tutorial (based on the auditory example in the SPM manual). I'm trying to make it sufficiently detailed so that an absolute AA newbie could successfully run the analysis. Perhaps we could include it as an appendix on the wiki.

**Reply by tiborauer** [reply]
Keeping the parametersets as they are means we consider them as part of the aa, rather than as examples. The fully specified default set is indeed part of aa, however, the site-specific instances (including yours) may not be. On the other hand, I cannot figure out a better place to keep them even when considering a local installation only. Most sites have a central space for software and tools to avoid redundancy, and the parameterset is usually stored there, as well.

**Reply by jones-michael-s** [reply]
Here's a thought: 

1) We could add  a "data" directory to the repo (at the same level as aa_modules, examples, etc)
2) move aap_parameters_defaults.xml to /data
3) move all the other parameter sets to a subdirectory under /examples

This would be consistent with aap_parameters_defaults.xml  being the master parameter "template" but the other parameter files currently in the distribution being examples of setting up a parameter file for a specific installation.

A data directory could also have other uses. For example, the diag (diagnostic) function in aamod_segment8 uses the MNI152 template from FSL. If you're not using any other FSL functionality, you have to install FSL _just to get this one file._ Instead, we could include the template in the data directory.

Yes, a somewhat silly example (we could change diag to use an SPM template) but its not hard to image other AA operations that would benefit from a data directory.

**Reply by tiborauer** [reply]
I like it, however, the question is still there: Where should the local maintainer/lab manager/user put their site-specific parameterset?

**Comment by jones-michael-s** [comment]
> Where should the local maintainer/lab manager/user put their site-specific parameterset?

Couldn't the manager simply give each user a copy of the local parameter file (i.e., as part of aa "training")? As long as it specifies a fullpath in the xi:include, the user's parameter file can live anywhere, yes?

Alternatively, this could be another use of a /data directory. I'm still not thrilled about anything going in the aa directory tree that isn't an "official" file but I guess I'm not the aa police...

**Reply by tiborauer** [reply]
The multitude of (potentially uncontrolled) copies of local parameterset files may lead to non-transparent fragmentation of the workflows and would make project management and support very inefficient. The manager would need to update the individual copies. I prefer a central copy and transparent customization of the parameters either in the user script or in project/user-specific copies of the parameterset.

**Reply by tiborauer** [reply]
I have seen other tools providing a configuration/parameterset file as part of the repo, where the user customise this file locally without committing to the repo. Our solution of specialization based on inheritance is better than that IMHO.

**Reply by jones-michael-s** [reply]
> I have seen other tools providing a configuration/parameterset file as part of the repo...

Couldn't a site manager put their parameter file someplace like /usr/local?

**Reply by tiborauer** [reply]
That is also possible, however, then it requires full path reference to the default parameterset, and the users have to add this folder to the MATLAB path or use full path reference to the site-specific parameterset when calling `aarecipe`. We can recommend that, too; and the site managers can decide themselves.

The bottom line is. that you should go ahead with your suggestion with the data folder, IMHO.

**Reply by jones-michael-s** [reply]
Should I include it in my current PR (for aa_test)?

**Reply by jones-michael-s** [reply]
(I think it will break things for people with parameter files in /aa_parametersets)

**Reply by tiborauer** [reply]
You may be right. If they use full path, then they should change it. If they use just the filenames, then although /data and /examples will be added to the path, the reference to the default parameterset will be invalid, and the XML inclusion will fail. They have to either move the site-specific parameterset to /data or edit it.

**Reply by tiborauer** [reply]
Maybe we should make a note and add the change in aa6.

**Reply by jones-michael-s** [reply]
Agreed
